lib: add option to silence warnings

PoojaDurgad opened the pull request:
Introduce an option `--suppress-warnings` to silence experimental and deprecation warnings for specified features

Fixes: https://github.com/nodejs/node/issues/30810

Co-Authored-By: Cody Deckard <cjdeckard@gmail.com>
<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

<!--
Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
-->

benjamingr commented:
Why?

cjihrig commented:
> Why?

There is prior discussion in https://github.com/nodejs/node/issues/30810 and https://github.com/nodejs/node/pull/31000.

benjamingr commented:
@cjihrig in that case this PR goes against the [TSC consensus discussed in the PR](https://github.com/nodejs/node/pull/31000#issuecomment-629811211)?

cjihrig commented:
I haven't reviewed this PR yet, but IIRC, the idea was that we could support silencing specific warnings, not just experimental warnings.

mscdex commented on lib/internal/util.js:
```diff
@@ -176,9 +176,15 @@ function slowCases(enc) {
+  }
   const msg = `${feature} is an experimental feature. This feature could ` +
-       'change at any time';
+  'change at any time';
```
I think left-aligning this with the beginning of the string on the previous line like this:

```suggestion
              'change at any time';
```

will read better and will look less like a mistake or some kind of `'use strict';`-like directive.

mscdex commented on doc/api/cli.md:
```diff
@@ -577,6 +577,13 @@ added: v0.8.0
 
 Silence deprecation warnings.
 
+### `--no-experimental-warnings`
```
```suggestion
### `--no-experimental-warnings=feature`
```

mscdex commented:
It looks like this option doesn't allow specifying multiple features, is this intentional?

PoojaDurgad commented:
I'm working to address the issues.

jasnell reviewed (changes requested):
Any `process.emitWarning()` may [include a `code` option](https://nodejs.org/dist/latest-v15.x/docs/api/process.html#process_process_emitwarning_warning_options).

Rather than introducing a new cli option just to disable experimental warnings, we should extend the existing `--no-warnings` to allow silencing specific codes:

```sh
$ node --no-warnings=EXP001,DEP002
```

Or, by warning type:

```sh
$ node --no-warnings=ExperimentalWarning,DeprecationWarning
```

Note: We do not currently assign individual codes to experimental warnings. We should start doing so.

Additionally, the way this PR makes the change, experimental warnings would not be emitted to the `process.on('warning')` event, which is a mistake. The existing `--no-warnings` flag will only suppress the default action of printing warnings to `stderr` but those will still be emitted to `process.on('warning')`. This is intentional behavior that should be preserved here as well.

PoojaDurgad commented:
@mscdex , @jasnell , @benjamingr - I pushed some  changes . please have a look.

- Introduced new flag `--suppress-warnings` 
- Introduced codes for experimental warnings.
- Enabled the flag to take multiple codes.
- Adjusted tests to the changes.
- Added a test.

jasnell commented on doc/api/cli.md:
```diff
@@ -848,6 +848,13 @@ Enables report to be generated on uncaught exceptions. Useful when inspecting
+added: REPLACEME
+-->
+
+Silence deprecation and experimental warnings for specified codes.
```
What's the difference between this and the existing --no-warnings flag?

PoojaDurgad replied:
@jasnell - I have introduced the new flag `--suppress-warnings` and also introduced codes for experimental warnings which enables this flag to silence the experimental and deprecation  warnings for supplied codes. The idea is to support silencing warnings for experimental and deprecation warnings and enabled the flag to take multiple codes which is not the case with existing `--no-warnings`

jasnell replied:
Yes, and what I'm saying is that the `--no-warnings` flag can just be extended to support this without introducing a new flag. `--no-warnings` (with no arguments) would continue to work as it does today, while `--no-warnings={list of codes}` would suppress the named codes. In either case, the documentation here should indicate that the flag takes a list of codes.

PoojaDurgad replied:
@jasnell - well instead of introducing new clip option we can extend the existing `--no-warnings` option right . yeah I understood and thanks for the review.

jasnell commented on lib/internal/util.js:
```diff
@@ -175,11 +175,20 @@ function slowCases(enc) {
+  });
+  if (found) return;
+
+  const msg = `[${code}] ${feature} is an experimental feature. ` +
```
This is not the right way to include the code. The emitWarning api already has an option for including the code correctly. Also, I would add this check into the emitWarning processing so it can be applied generically to all warning types.

jasnell replied:
Not sure why this comment was marked resolved as the issue is still there.

PoojaDurgad replied:
without knowing,  it was marked as resolved

PoojaDurgad replied:
@jasnell - I have a doubt , can you confirm one thing please , the experimental warning `vm.measureMemory` API calls emitExperimentalWarning like this `emitExperimentalWarning('vm.measureMemory');`  https://github.com/nodejs/node/blob/22293eab481548c7dba3ffd320487d4f267b977b/lib/vm.js#L390 so the function takes a string and it matches with the `function emitExperimentalWarning(feature)`  like this https://github.com/nodejs/node/blob/51b43675067fafaad0abd7d4f62a6a5097db5044/lib/internal/util.js#L188 so if I'm not wrong this call does not include an option for codes. If the above file modification is not the right way to include the code, then how to do that? can you suggest me.

bnoordhuis commented:
I'm going to close this because the pull request has a great many conflicts and because the requested feature from #30810 has been implemented but thanks anway, @PoojaDurgad, it's much appreciated.

haidaqn commented:
**I am getting this error when integrating nextjs 13 with clerk and using node 16.13.0** : "  ExperimentalWarning: stream/web is an experimental feature. This feature could change at any time
(Use `node --trace-warnings ...` to show where the warning was created) "
